Updated export features with new backend

declanblanc opened the pull request:
## **Description**: 

Apologies for the poorly named branch, I started working on this alongside the backend branch `feat-bulk-crns` and along the way ended up reimplementing the export features as a means to test the backend API.

This branch is primarily just copy/paste from feat-export links with some small tweaks. 
Important to note testing this branch requires backend branch `feat-bulk-crns` as well.

## **Changes Made**:

Created `getSectionDataByCrn.tx` server action. It is very similar to getSectionData.ts however this function takes the frontend URL as a string and uses that same URL to communicate with the API. Rather than returning info for a single course, it returns course info for each CRN that was passed.

Updated `Shell.tsx` to handle query parameters when users navigate to the website using a sharable link.

Updated `PlanMenu.tsx` with buttons and functions for each export feature (link, txt, & csv)

Updated `planStore.ts` to give the “Plan” object an optional “isTemporary” boolean. This allows the plan preview provided by the export link. e.g. When a user follows a link that was sent to them to see someone else’s plan, the plan is not added to their local list of plans unless they decide it should be.

## **Testing Steps**:

Very similar to the original testing steps of feat-export-link:

Tested all export methods with a variety of plans including:

- All in-person
- All online
- Large mix of in-person and online
- Mixture of courses with and without selected sections

Previously, the export link would become rather buggy with any more than 18 courses in a plan. This no longer seems to be the case from all my testing. Shoutout NestJS and Mongoose I suppose!

declanblanc committed:
added urlSave to share menu, updated plans to include isTemporary boolean for previewing plans via link share. implementing getSectionDataByCrns

declanblanc committed:
Export features now updated for compatability with dev branch and NestJS backend. This commit is using the backend branch 'feat-declan-crn' for db queries.

declanblanc committed:
shell

Alejandro-Gonzale-z reviewed (changes requested):
The plan doesn't save properly if you are logged in. 
Can we add an indicator that says hey you are temporarily viewing this plan or something along this lines so that when users open a plan they know they're only viewing not autosaving.  

CSS Changes
The save plan notification should not be red, red = something bad happened.
The share buttons should all be the same width & height.

WokFriedE committed:
Merge branch 'dev' into feat-bulk-crns

declanblanc committed:
unstable commit for ethan to help

declanblanc committed:
temp plan improvements

declanblanc committed:
header

declanblanc committed:
issues with temporary plans persisting and uploading to db

Alejandro-Gonzale-z committed:
fixed auto sync for temp plans

declanblanc committed:
Share link functioning properly. Added QOL features for temporary plans to plan menu

declanblanc committed:
Moved json share outside of share and to its own export button

Alejandro-Gonzale-z committed:
Merge branch 'dev' into feat-bulk-crns

Alejandro-Gonzale-z committed:
fixed merge conflict issues

WokFriedE reviewed (changes requested):
lgtm and works great awesome job - check out the nitpick other comments

declanblanc committed:
pushing solely for the sake of sharing

this is nasty rn, errors all over the place. just pushing bc I can't work on it.

but it does run locally and allows you to share links :D
